pkg/handler/collector/gcs/gcs_test.go

```diff
@@ -0,0 +1,140 @@
+// limitations under the License.
+
+package gcs
+
+import (
+	"context"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/fsouza/fake-gcs-server/fakestorage"
+	"github.com/guacsec/guac/pkg/handler/processor"
+)
+
+func TestGCS_RetrieveArtifacts(t *testing.T) {
+	os.Setenv("GCS_BUCKET_ADDRESS", "some-bucket")
+	ctx := context.Background()
+	server := fakestorage.NewServer([]fakestorage.Object{
+		{
+			ObjectAttrs: fakestorage.ObjectAttrs{
+				BucketName: "some-bucket",
+				Name:       "some/object/file.txt",
+				Updated:    time.Date(2009, 11, 17, 20, 34, 58, 651387237, time.UTC),
+			},
+			Content: []byte("inside the file"),
+		},
+	})
+	defer server.Stop()
+	client := server.Client()
+
+	var doc *processor.Document = &processor.Document{
+		Blob: []byte("inside the file"),
+		SourceInformation: processor.SourceInformation{
+			Collector: string(CollectorGCS),
+			Source:    getBucketPath(),
+		},
+	}
+
+	type fields struct {
+		bucket       string
+		reader       gcsReader
+		lastDownload time.Time
+		poll         bool
+	}
+	tests := []struct {
+		name     string
+		fields   fields
+		want     *processor.Document
+		wantErr  bool
+		wantDone bool
+	}{
+		{
+			name:    "no reader",
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name: "get object",
+			fields: fields{
+				bucket: getBucketPath(),
+				reader: &reader{client: client, bucket: getBucketPath()},
+			},
+			want:     doc,
+			wantErr:  false,
+			wantDone: true,
+		},
+		{
+			name: "last download time the same",
+			fields: fields{
+				bucket:       getBucketPath(),
+				reader:       &reader{client: client, bucket: getBucketPath()},
+				lastDownload: time.Date(2009, 11, 17, 20, 34, 58, 651387237, time.UTC),
+			},
+			want:     nil,
+			wantErr:  false,
+			wantDone: true,
+		},
+		{
+			name: "last download time set before",
+			fields: fields{
+				bucket:       getBucketPath(),
+				reader:       &reader{client: client, bucket: getBucketPath()},
+				lastDownload: time.Date(2009, 10, 17, 20, 34, 58, 651387237, time.UTC),
+			},
+			want:     doc,
+			wantErr:  false,
+			wantDone: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := &gcs{
+				bucket:       tt.fields.bucket,
+				reader:       tt.fields.reader,
+				lastDownload: tt.fields.lastDownload,
+				poll:         tt.fields.poll,
+			}
+			docChan := make(chan *processor.Document, 1)
+			errChan := make(chan error, 1)
+			go func() {
+				errChan <- g.RetrieveArtifacts(ctx, docChan)
+			}()
+			numCollectors := 1
+			collectorsDone := 0
+			for collectorsDone < numCollectors {
+				select {
+				case d := <-docChan:
+					if !reflect.DeepEqual(d, tt.want) {
+						t.Errorf("g.RetrieveArtifacts() = %v, want %v", d, tt.want)
+					}
+				case err := <-errChan:
+					if (err != nil) != tt.wantErr {
+						t.Errorf("g.RetrieveArtifacts() error = %v, wantErr %v", err, tt.wantErr)
+						return
+					}
+					collectorsDone += 1
+				}
+			}
```

Comment:
there is a potential race here, still need to drain the rest of the docChan

Reply:
fixed.